### ISSUE TITLE
Using the same class for 2 different models (code first) throws random exceptions

### ISSUE THREAD

**Issue by paya-cz** [dotnet/efcore#9985]
I have a single POCO model class (without any attribute decoration), and a single `DatabaseContext` class. The class is using code-first model building, and importantly it's using `if` statements to build the model in `OnModelCreating`, dynamically including/excluding properties from the POCO class based on some value passed to the `DatabaseContext` in the constructor. The program works as follows:

Create `DatabaseContext` with some input value, resulting in model version 1 of the POCO class.
Use the database context, connect to a database, everything works fine.
Dispose the database context.
Now, create another `DatabaseContext` with a different constructor parameter, resulting in a model version 2 of the same POCO class.
Now, trying to use the class to deserialize stored proc result throws a pretty much random exception. At first, it was throwing `IndexOutOfRangeException` (see below). When I reversed the model versions (eg first creating and using the version 2, then version 1), it started throwing `InvalidCastException` instead, in a completely different location.

```
System.IndexOutOfRangeException: Index was outside the bounds of the array. 
at Microsoft.EntityFrameworkCore.Storage.Internal.RemappingUntypedRelationalValueBufferFactory.Create(DbDataReader dataReader) 
at Microsoft.EntityFrameworkCore.Query.Internal.QueryingEnumerable.Enumerator.MoveNext() 
at Microsoft.EntityFrameworkCore.Query.QueryMethodProvider.<_ShapedQuery>d__3`1.MoveNext() 
at System.Linq.Enumerable.WhereEnumerableIterator`1.MoveNext() 
at System.Linq.Enumerable.<TakeIterator>d__197`1.MoveNext() 
at Microsoft.EntityFrameworkCore.Query.Internal.LinqOperatorProvider.ExceptionInterceptor`1.EnumeratorExceptionInterceptor.MoveNext() 
at System.Collections.Generic.EnumerableHelpers.ToArray[T](IEnumerable`1 source, Int32& length) 
at System.Collections.Generic.EnumerableHelpers.ToArray[T](IEnumerable`1 source) 


System.InvalidCastException: Unable to cast object of type 'System.String' to type 'System.Nullable`1[System.Int32]'. 
at lambda_method(Closure , ValueBuffer ) 
at Microsoft.EntityFrameworkCore.Query.ExpressionVisitors.Internal.UnbufferedEntityShaper`1.Shape(QueryContext queryContext, ValueBuffer valueBuffer) 
at Microsoft.EntityFrameworkCore.Query.QueryMethodProvider.<_ShapedQuery>d__3`1.MoveNext() 
at System.Linq.Enumerable.WhereEnumerableIterator`1.MoveNext() 
at System.Linq.Enumerable.<TakeIterator>d__197`1.MoveNext() 
at Microsoft.EntityFrameworkCore.Query.Internal.LinqOperatorProvider.ExceptionInterceptor`1.EnumeratorExceptionInterceptor.MoveNext() 
at System.Collections.Generic.EnumerableHelpers.ToArray[T](IEnumerable`1 source, Int32& length) 
at System.Collections.Generic.EnumerableHelpers.ToArray[T](IEnumerable`1 source) 
```

### Further technical details
EF Core version: 1.0.5
Database Provider: Microsoft.EntityFrameworkCore.SqlServer
Operating system: Win7
IDE: Visual Studio 2017

**Comment by smitpatel** [comment]
This is expected since model is cached.
EFCore runs `OnModelCreating` only once and cache the model created to be used subsequently. If you are doing conditional statements in your `OnModelCreating` then you are effectively creating multiple models for same type of derived DbContext. In such case, you need to replace `ModelCacheKeyFactory` service in service provider to use your own, which would generate a `ModelCacheKey` which differentiate based on the property you are passing in constructor. (Effectively generating different CacheKey based on different values of property.) There will be one model in service provider corresponding to each CacheKey.

**Comment by paya-cz** [comment]
@smitpatel Thank you, that makes a lot of sense now. Is there any documentation on how to use `ModelCacheKeyFactory`?

**Comment by smitpatel** [comment]
@paya-cz - There is no documentation at present. There is general idea to improve such scenarios without needing users to override services. It is being tracked here #9965 

For docs, you can file issue at docs repo https://github.com/aspnet/EntityFramework.Docs/issues

**Comment by paya-cz** [comment]
@smitpatel Thanks, I have added a new issue to improve the documentation.
